src/tir/schedule/primitive/loop_transformation.cc

```diff
@@ -384,6 +510,182 @@ StmtSRef Fuse(ScheduleState self, const Array<StmtSRef>& loop_srefs) {
+                                LoopsNotAChainError::ProblemKind::kHaveNonSingleBranchStmt);
+    }
+    chain.push_back(loop_sref);
+    loop_sref = parent_loop_sref;
+  }
+  chain.push_back(top);
+  return chain;
+}
+
+/*!
+ * \brief Construct a loop chain in the new order
+ * \param self The schedule state
+ * \param chain The loops in the reorder range
+ * \param ordered_loop_srefs The loop srefs to be reordered
+ * \param loop_srefs The set containing loop srefs to be reordered
+ * \return the new loop chain
+ * \throws ScheduleError If the domain of an outer loop depends on any of the inner loops after
+ * reordering
+ */
+For ConstructNewLoopChain(const ScheduleState& self, std::vector<const StmtSRefNode*> chain,
+                          const Array<StmtSRef>& ordered_loop_srefs,
+                          const std::unordered_set<const StmtSRefNode*>& loop_srefs) {
+  std::unordered_set<const VarNode*> inner_vars;
+  inner_vars.reserve(chain.size());
+  For new_loop{nullptr};
+  int index = static_cast<int>(ordered_loop_srefs.size()) - 1;
+  for (const StmtSRefNode* loop_sref : chain) {
+    const ForNode* copy = nullptr;
+    if (loop_srefs.count(loop_sref)) {
+      copy = ordered_loop_srefs[index]->StmtAs<ForNode>();
+      --index;
+    } else {
+      copy = loop_sref->StmtAs<ForNode>();
+    }
+    ICHECK(copy != nullptr);
+    ObjectPtr<ForNode> n = make_object<ForNode>(*copy);
+    if (new_loop.defined()) {
+      n->body = new_loop;
+    } else {
+      n->body = loop_sref->StmtAs<ForNode>()->body;
+    }
+    const VarNode* used_var = nullptr;
+    auto f_contain = [&inner_vars, &used_var](const VarNode* var) {
+      if (inner_vars.count(var)) {
+        used_var = var;
+        return true;
+      }
+      return false;
+    };
+    if (UsesVar(copy->min, f_contain) || UsesVar(copy->extent, f_contain)) {
+      throw DependentLoopError(self->mod, GetRef<For>(copy), used_var->name_hint);
+    }
+    inner_vars.insert(copy->loop_var.get());
+    new_loop = For(std::move(n));
+  }
+  return new_loop;
+}
+
+void Reorder(ScheduleState self, const Array<StmtSRef>& ordered_loop_srefs) {
+  if (ordered_loop_srefs.size() <= 1) {
+    return;
+  }
+  // Step 1. Check uniqueness.and collect the input loop srefs into a set
+  std::unordered_set<const StmtSRefNode*> loop_srefs =
+      CollectLoopsIntoSet(self, ordered_loop_srefs);
+  // Step 2. Gather loops to be reordered
+  // For each loop sref in the input sref array, traverse upwards along its parent pointer in the
+  // sref tree, and stop on either a block, or a previously-visited loop
+  // - the top of the reorder range is the last loop visited in the first traversal which exists in
+  //   the input array
+  // - the bottom of the reorder range is the last loop in the input array which is not visited in
+  // the previous traversals
+  auto pair = GetBoundaryOfReorderRange(self, loop_srefs);
+  const StmtSRefNode* top = pair.first;
+  const StmtSRefNode* bottom = pair.second;
```

Comment:
In this way we can avoid creating a temporary pair.
```suggestion
  const StmtSRefNode* top;
  const StmtSRefNode* bottom;
  std::tie(top, bottom) = GetBoundaryOfReorderRange(self, loop_srefs);
```